tests/backend_test.py

```diff
@@ -1430,3 +1432,63 @@ def test_ibmq_local_emulator(
     counts = r.get_counts()
     # Most results should be (0,0) or (1,1):
     assert sum(c0 != c1 for c0, c1 in counts) < 25
+
+
+# https://github.com/CQCL/pytket-qiskit/issues/231
+def test_noiseless_density_matrix_simulation() -> None:
+    density_matrix_backend = AerDensityMatrixBackend()
+    assert density_matrix_backend.supports_density_matrix is True
+
+    assert isinstance(density_matrix_backend.backend_info.architecture, FullyConnected)
+
+    circ1 = Circuit(3).X(0).X(1).CCX(0, 1, 2)
+
+    output_state = np.array([0] * 7 + [1])
+
+    result1 = density_matrix_backend.run_circuit(circ1)
+
+    assert result1.get_density_matrix().shape == (8, 8)
+    assert np.allclose(
+        result1.get_density_matrix(), np.outer(output_state, output_state.conj())
+    )
+    # Example with resets and conditional gates
+    # Deterministic if input is a computational basis state
+    circ2 = Circuit(3, 1)
+    circ2.CCX(*range(3))
+    circ2.U1(1 / 4, 2)
+    circ2.H(2)
+    circ2.Measure(2, 0)
+    circ2.CZ(0, 1, condition_bits=[0], condition_value=1)
+    circ2.Reset(2)
+
+    result2 = density_matrix_backend.run_circuit(circ2)
+    assert result1.get_density_matrix().shape == (8, 8)
+    state_backend = AerStateBackend()
+    statevector = state_backend.run_circuit(circ2).get_state()
+    assert np.allclose(
+        result2.get_density_matrix(), np.outer(statevector, statevector.conj())
+    )
+
+
+# https://github.com/CQCL/pytket-qiskit/issues/231
+def test_noisy_density_matrix_simulation() -> None:
+
+    # Test that __init__ works with a very simple noise model
+    noise_model = NoiseModel()
+    noise_model.add_quantum_error(depolarizing_error(0.6, 2), ["cz"], [0, 1])
+    noise_model.add_quantum_error(depolarizing_error(0.6, 2), ["cz"], [1, 2])
+
+    noisy_density_sim = AerDensityMatrixBackend(noise_model)
+    assert isinstance(noisy_density_sim.backend_info.architecture, Architecture)
+    assert len(noisy_density_sim.backend_info.architecture.nodes) == 3
+
+    circ = Circuit(3)
+    circ.X(0)
+    circ.X(1)
+    circ.SX(1)
+    circ.CZ(0, 1)
+    circ.CZ(1, 2)
+    assert noisy_density_sim.valid_circuit(circ)
+
+    result = noisy_density_sim.run_circuit(circ)
+    assert result.get_density_matrix().shape == (8, 8)
```

Comment:
Can we also check that the density matrix differs from what we get with a non-noisy simulation?

Reply:
The way I've done this so far is to check the purity of the state represented by the density matrix

$$
\gamma = \text{Tr}(\rho^2)
$$

If $\gamma =1$, the state is pure. If $\gamma < 1$ the state is mixed. I've verified in [6d00b85](https://github.com/CQCL/pytket-qiskit/pull/380/commits/6d00b85ca6cf7e90cc5cec04aa21bcac566672c5) that the density matrix given by the noisy simulator is a mixed state and the noisless a pure state.

I'm not sure whether this is thorough enough. Perhaps its a good idea to calcluate the expectation value of some observable $E$ and verify that they are not equal.